### PR TITLE
refactor: split code

### PR DESCRIPTION
**Pull request by marionebl**
follow-up for #46

### CODE REVIEW

**Comment by sindresorhus** [comment]
Can you fix the merge conflict?

Also make sure you take into account: https://github.com/sindresorhus/atom-linter-xo/commit/e8209a871bb21fb8f281e1487e43c8a4a215edd2 https://github.com/sindresorhus/atom-linter-xo/commit/8194093d97e38012be576a699325f2c98ca6a59c https://github.com/sindresorhus/atom-linter-xo/commit/a9661b28973a02ed6f7ccb7caea3c4d262559122

File: index.js [hunk 1]
```diff
@@ -151,35 +23,48 @@ export function activate() {
 				return;
 			}
 
-			fix(editor);
+			const text = editor.getText();
+			fix(editor)(text);
```
Comment (sindresorhus) [inline index.js]:
```js
fix(editor)(editor.getText());
```

File: index.js [hunk 2]
```diff
@@ -151,35 +23,48 @@ export function activate() {
...
+	this.subscriptions.dispose();
+}
+
+export function provideLinter() {
+	return {
+		name: 'XO',
+		grammarScopes: SUPPORTED_SCOPES,
+		scope: 'file',
+		lintOnFly: true,
+		lint: editor => lint(editor)(editor.getText()).then(format(editor))
```
Comment (sindresorhus) [inline index.js]: I usually only do inline arrow functions for things that return something:

```js
lint: editor => {
    lint(editor)(editor.getText()).then(format(editor));
}
```

File: lib/fix.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+/** @babel */
+import lint from './lint';
+
+// (editor: Object) => function
```
Comment (sindresorhus) [inline lib/fix.js]: What's the purpose of these notations?

File: lib/get-xo.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+/** @babel */
+import {allowUnsafeNewFunction} from 'loophole';
+import getPackagePath from './get-package-path';
+
+// (base: string) => xo: Promise<Object>
+export default function getXO(base) {
+	const getResolvePath = () => base ? getPackagePath(base) : Promise.resolve();
+
+	return getResolvePath()
```
Comment (sindresorhus) [inline lib/get-xo.js]: Why make this async? It's only a few fs calls, and the sync version would be much faster.

File: lib/fix.js (hunk 3 above)
Comment (marionebl) [inline lib/fix.js]: A habit I picked up when working on projects without type system. Remove or convert them to full-fledged JsDoc?

Comment (sindresorhus) [inline lib/fix.js]: Ok. Was just curious. It's fine like this. Makes the code more readable. At some point I'll probably convert the codebase to Flow type.

File: lib/get-xo.js (hunk 4 above)
Comment (marionebl) [inline lib/get-xo.js]: I'm not sure on how to obtain the benchmarks. I'll just make this sync, ok?

Comment (marionebl) [inline lib/get-xo.js]: `getXO` needs `getPackagePath`, which actually reads and parses `package.json` files. Before changing all the underlying utils to `sync` counterparts I'd rather do the benchmarks. Do you have an idea on how to go about this?

**Comment by marionebl** [comment]
* **XO:Fix**: https://github.com/sindresorhus/atom-linter-xo/commit/8194093d97e38012be576a699325f2c98ca6a59c - included
* **Directory null check**: https://github.com/sindresorhus/atom-linter-xo/commit/a9661b28973a02ed6f7ccb7caea3c4d262559122 - superseeded

---
* **Smarter Ranges**: https://github.com/sindresorhus/atom-linter-xo/commit/e8209a871bb21fb8f281e1487e43c8a4a215edd2 - This one is more tricky. I moved the logic to derive a `Range` from the eslint `pos` attribute to `selectMessageRange` ([source](https://github.com/sindresorhus/atom-linter-xo/pull/47/files#diff-841daed463ab7befab0f2d50bf758bf9R58)). This includes usage of `endLine`, and `endColumn`. Apparently `rangeFromLineNumber` does not support this. Which variant to pick?

**Comment by marionebl** [comment]
@sindresorhus Could you have another look at answer the questions on https://github.com/sindresorhus/atom-linter-xo/pull/47#issuecomment-262248069 and https://github.com/sindresorhus/atom-linter-xo/pull/47#discussion_r89118337? Thanks!

File: lib/get-xo.js (hunk 4 above)
Comment (sindresorhus) [inline lib/get-xo.js]: We can leave it async for now and do some benchmarking when all the pending changes are done. Mind opening an issue about benchmarking this so we don't forget?

Comment (marionebl) [inline lib/get-xo.js]: Done! https://github.com/sindresorhus/atom-linter-xo/issues/52

File: index.js [hunk 5]
```diff
@@ -151,40 +23,49 @@ export function activate() {
...
+				const {scopeName} = editor.getGrammar();
 
-				if (isJS && shouldFixOnSave && dependsOnXO) {
-					fix(editor);
+				if (!SUPPORTED_SCOPES.includes(scopeName)) {
+					return;
 				}
+
+				const text = editor.getText();
+				return fix(editor)(text);
```
Comment (sindresorhus) [inline index.js]:
```js
return fix(editor)(editor.getText());
```

**Comment by sindresorhus** [comment]
> This one is more tricky. I moved the logic to derive a Range from the eslint pos attribute to selectMessageRange (source). This includes usage of endLine, and endColumn. Apparently rangeFromLineNumber does not support this. Which variant to pick?

What's the benefit of your vs the other one and inverse? It's a bit unclear to me what the practical differences are. There must be a reason `atom-linter-eslint` uses `rangeFromLineNumber`. The other method also improves the error when a rule returns an invalid range. // @jonathandelgado

@steelbrain Why can't Atom Linter just use `rangeFromLineNumber` internally, instead of us having to do so much [boilerplate](https://github.com/sindresorhus/atom-linter-xo/commit/e8209a871bb21fb8f281e1487e43c8a4a215edd2)?

**Comment by marionebl** [comment]
The benefit of the approach using the `end*` attributes is the ability to mark multiple lines with a single problem (e.g. for [no-multiple-empty-lines](http://eslint.org/docs/rules/no-multiple-empty-lines)).

**rangeFromLineNumber**: https://github.com/sindresorhus/atom-linter-xo/commit/e8209a871bb21fb8f281e1487e43c8a4a215edd2:

bildschirmfoto 2016-11-23 um 09 12 42


**end\***: https://github.com/sindresorhus/atom-linter-xo/pull/47/commits/ab2328644c623f6b1081490e01cad03a5b249e1b

bildschirmfoto 2016-11-23 um 09 20 57

File: index.js (hunk 2 above)
Comment (marionebl) [inline index.js]: provider.lint actually has to return the Array<Message>

**Comment by sindresorhus** [comment]
Ah, that's nice. How does your version handle when there's no range information? The reason the `rangeFromLineNumber` was added, is so that it would show underline for methods/words when there's no range information. Maybe we could fall back to `rangeFromLineNumber` when there's no range info?

**Comment by marionebl** [comment]
Fused the approaches

File: lib/format.js [hunk 6]
```diff
@@ -12,13 +13,18 @@ export default function format(editor) {
 		if (!report) {
 			return [];
 		}
+
 		const [{messages}] = report.results;
+
 		return messages.map(message => {
+			const [err, range = []] = withErrorHandling(selectMessageRange)(editor, message);
```
Comment (sindresorhus) [inline lib/format.js]: We should not handle the error or display it nicely. It's a severe edge-case and should cause it to fail.

**Comment by marionebl** [comment]
Changed as requested

File: lib/format.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+	const end = [endLine, endColumn]
+		.filter(coord => typeof coord === 'number')
+		.map(coord => coord - 1);
+
+	if (end.length === 2) {
+		return [start, end];
+	}
+
+	try {
```
Comment (sindresorhus) [inline lib/format.js]: This is never reached as `end.length` is always `2`.

Master:
screen shot 2016-11-23 at 22 19 32

This PR:
screen shot 2016-11-23 at 22 20 01

Notice the underline.

File: lib/format.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+		return rangeFromLineNumber(editor, start[0], start[1]);
```
Comment (jonathandelgado) [inline lib/format.js]: If `message.endLine` and `message.endColumn` are provided, we should not run `rangeFromLineNumber` - we should simply return a new range, see [the current implementation](https://github.com/sindresorhus/atom-linter-xo/blob/e8209a871bb21fb8f281e1487e43c8a4a215edd2/index.js#L79)

I think the maps/filters add unnecessary complexity, as well.

**Comment by jonathandelgado** [comment]
> What's the benefit of your vs the other one and inverse? It's a bit unclear to me what the practical differences are. There must be a reason atom-linter-eslint uses rangeFromLineNumber. The other method also improves the error when a rule returns an invalid range. // @jonathandelgado

I reviewed the most recent changes for ranges. To note, the current implementation uses the `endLine` and `endColumn` [here](https://github.com/sindresorhus/atom-linter-xo/blob/e8209a871bb21fb8f281e1487e43c8a4a215edd2/index.js#L79), if there is not one provided, it will try to assume the range using `rangeFromLineNumber`. This is extremely important for an accurate range.

I'll make some inline comments right now on 

> The benefit of the approach using the end* attributes is the ability to mark multiple lines with a single problem (e.g. for no-multiple-empty-lines).

I agree that the multiple lines issue is a bug; this should be fixed in `atom-linter` though, so all the linters can benefit.

**Comment by marionebl** [comment]
I reverted `selectMessageRange` to the current inline implementation, this should produce the exact same behavior as currently exhibited by atom-linter.

**Comment by sindresorhus** [comment]
Yes. This is excellent. Really appreciate your work on improving this plugin @marionebl :)

superhighfive

**Comment by sindresorhus** [comment]
@marionebl Would you be interested in joining the [XO](https://github.com/sindresorhus/xo) project? You produce very high quality code and seem interested in the project. No worries if not though :)

**Comment by marionebl** [comment]
@sindresorhus Sure, XO is great 🎉 Thanks for the nice feedback, appreciate it.

**Pull request merged**